apptools/preferences/preferences_helper.py

```diff
@@ -65,9 +72,23 @@ def _preferences_default(self):
     def _anytrait_changed(self, trait_name, old, new):
         """ Static trait change handler. """
 
+        if self.preferences is None:
+            return
+
+        # If the trait was a list or dict '_items' trait then just treat it as
+        # if the entire list or dict was changed.
+        if trait_name.endswith('_items'):
+            trait_name = trait_name[:-6]
+            if self._is_preference_trait(trait_name):
+                self.preferences.set(
+                    '%s.%s' % (self._get_path(), trait_name),
+                    getattr(self, trait_name)
+                )
+                return
+
         # If we were the one that set the trait (because the underlying
         # preferences node changed) then do nothing.
-        if self.preferences and self._is_preference_trait(trait_name):
+        if self._is_preference_trait(trait_name):
```

Comment:
This is likely an unrealistic edge case (it would probably be considered a programmer error), but what if there is a preference trait that ends with `_items`, e.g. `something_items`?  In that case we enter the above if statement, and set `trait_name = = trait_name[:-6]` removing the trailing `_items` leaving us with `something`.  Then calling `getattr(self, trait_name)` will fail.  

In general you should never name traits with a trailing `_items`, so this code is good to stay as is. But just wanted to ask about this.
e.g. this test fails:
<details>

```
def test_mutate_list_of_values(self):
        """ Mutated list should be saved and _items events not to be
        saved in the preferences.
        """
        # Regression test for enthought/apptools#129

        class MyPreferencesHelper(PreferencesHelper):
            preferences_path = Str('my_section')

            list_of_str_items = List(Str)

        helper = MyPreferencesHelper(list_of_str_items=["1"])

        # Now modify the list to fire _items event
        helper.list_of_str_items.append("2")
        self.preferences.save(self.tmpfile)

        new_preferences = Preferences()
        new_preferences.load(self.tmpfile)

        self.assertEqual(
            new_preferences.get("my_section.list_of_str_items"), str(["1", "2"])
        )
        self.assertEqual(new_preferences.keys("my_section"), ["list_of_str_items"])

```

</details>

Reply:
Yep, it is a very good observation. I agree this is why magic naming is bad, and one of the reasons to introduce `observe` is to do away with `_items` magic naming.

Reply:
I think the test you provided can be fixed though, but it may have to be done along with `apptools.preferences.ui.preferences_page`. I will do that in a separate PR.

To be honest, I think a better design would be one that does not require synchronization - we probably don't need that. Most of the time we care about storing the preferences to a file after it has been edited. As long as we can obtain and serialize all the values at the point when the file is saved, we are good. If someone says they want to share the same preferences with multiple helpers, I'd say, think again. A lot can go wrong when editing shared mutable states, best approach is don't.

Reply:
> I think the test you provided can be fixed though, but it may have to be done along with apptools.preferences.ui.preferences_page. I will do that in a separate PR.

Actually thinking about it... I am not sure it can be fixed to satisify both situations. The suffix `_items` is simply overloaded. It might be one that just needs to be documented and a warning be emitted perhaps.

Reply:
Opened https://github.com/enthought/apptools/issues/225